File: 102. Binary Tree Level Order Traversal/levelOrder.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct TreeNode
{
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int val, TreeNode* left = NULL, TreeNode* right = NULL)
		: val(val)
		, left(left)
		, right(right){}
	~TreeNode() { if (left) delete left; if (right) delete right; }
};

class Solution {
public:
	vector<vector<int> > levelOrder(TreeNode* root)
	{
		int ht = height(root);
		vector<vector<int> > ret;
		for (int i = 1; i <= ht; i++)
		{
			vector<int> level;
			PrintLevel(root, i, level);
			ret.push_back(level);
		}
		return ret;
	}

private:
	int height(TreeNode* root)
	{
		if (!root) { return 0; }
		return 1 + max(height(root->left), height(root->right));
	}

	int max(int a, int b) { return a > b ? a : b; }

	void PrintLevel(TreeNode* root, int level, vector<int>& ret)
	{
		if (!root) { return; }
		if (level == 1) { ret.push_back(root->val); return; }
		PrintLevel(root->left, level - 1, ret);
		PrintLevel(root->right, level - 1, ret);
	}
};

int main(void)
{
	TreeNode* root = new TreeNode(3, 
		new TreeNode(9, NULL, NULL),
		new TreeNode(20, 
				new TreeNode(15, NULL, NULL),
				new TreeNode(7, NULL, NULL)));
	Solution s;
	vector<vector<int> > ret = s.levelOrder(root);
	for (int i = 0; i < ret.size(); i++)
	{
		for (int j = 0; j < ret[i].size(); j++)
		{
			cout << ret[i][j] << " ";
		}
		cout << endl;
	}
	delete root;
	return 0;
}

File: 102. Binary Tree Level Order Traversal/levelOrder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private void levelOrder(List<Integer> result, TreeNode node, int level) {
        if (node == null) {
            return;
        }
        
        if (level == 1) {
            result.add(node.val);
            return;
        }
        levelOrder(result, node.left, level - 1);
        levelOrder(result, node.right, level - 1);
    }

    private int level(TreeNode root, int count) {
        if (root == null) {
            return count;
        }
        return Math.max(level(root.left, count + 1), level(root.right, count + 1));
    }

    public List<List<Integer>> levelOrder(TreeNode root) {
        // List<List<Integer>> result = new ArrayList<>();
        // int level = level(root, 0);
        // for (int i = 1; i <= level; i++) {
        //     List<Integer> levelResult = new ArrayList<>();
        //     result.add(levelResult);
        //     levelOrder(levelResult, root, i);
        // }
        // return result;
        return levelOrder2(root);
    }

    private void levelOrder2(List<List<Integer>> result, TreeNode node, int level) {
        if (node == null) {
            return;
        }

        if (level < result.size() + 1) {
            result.get(level - 1).add(node.val);
        } else {
            List<Integer> levelResult = new ArrayList<>();
            levelResult.add(node.val);
            result.add(levelResult);
        }

        levelOrder2(result, node.left, level + 1);
        levelOrder2(result, node.right, level + 1);
    }

    public List<List<Integer>> levelOrder2(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        levelOrder2(result, root, 1);
        return result;
    }
}